lte/gateway/c/oai/lib/pipelined_client/PipelinedServiceClient.cpp

```diff
@@ -0,0 +1,713 @@
+#include <cstring>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <thread>
+
+#include <grpcpp/impl/codegen/async_unary_call.h>
+
+#include "lte/protos/pipelined.grpc.pb.h"
+#include "lte/protos/pipelined.pb.h"
+#include "lte/protos/mobilityd.pb.h"
+#include "orc8r/protos/common.pb.h"
+#include "lte/protos/subscriberdb.pb.h"
+#include "ServiceRegistrySingleton.h"
+#include "common_defs.h"
+#include "proto_converters.h"
+
+namespace grpc {
+class Channel;
+class ClientContext;
+class Status;
+}  // namespace grpc
+
+using grpc::Channel;
+using grpc::ChannelCredentials;
+using grpc::ClientContext;
+using grpc::InsecureChannelCredentials;
+using grpc::Status;
+
+namespace magma {
+namespace lte {
+
+PipelinedServiceClient& PipelinedServiceClient::get_instance() {
+  static PipelinedServiceClient client_instance;
+  return client_instance;
+}
+
+PipelinedServiceClient::PipelinedServiceClient() {
+  auto channel = ServiceRegistrySingleton::Instance()->GetGrpcChannel(
+      "pipelined", ServiceRegistrySingleton::LOCAL);
+  stub_ = Pipelined::NewStub(channel);
+  std::thread resp_loop_thread([&]() { rpc_response_loop(); });
+  resp_loop_thread.detach();
+}
+
+//------------------- TUNNEL ADD -------------------
+
+//                    ADD : v4
+//--------------------------------------------------
+int PipelinedServiceClient::UpdateUEIPv4SessionSet(
+    const struct in_addr& ue_ipv4_addr, int vlan, struct in_addr& enb_ipv4_addr,
+    uint32_t in_teid, uint32_t out_teid, const std::string& imsi,
+    uint32_t flow_precedence, const std::string& apn, uint32_t ue_state,
+    std::function<void(Status, UESessionContextResponse)> callback) {
+  UESessionSet request = UESessionSet();
+  UESessionContextResponse response;
+
+  PipelinedServiceClient& client = get_instance();
+
+  // Set the UE IPv4 address
+  set_ue_ipv4_addr(ue_ipv4_addr, request);
+
+  // Set the vlan
+  request.set_vlan(vlan);
+
+  // Set the enb IPv4 address
+  set_gnb_ipv4_addr(enb_ipv4_addr, request);
+
+  // Set the incoming and outgoing teid
+  request.set_in_teid(in_teid);
+  request.set_out_teid(out_teid);
+
+  // Set the Subscriber ID
+  SubscriberID* sid = request.mutable_subscriber_id();
+  sid->set_id("IMSI" + std::string(imsi));
+  sid->set_type(SubscriberID::IMSI);
+
+  // Set the precedence
+  request.set_precedence(flow_precedence);
+
+  // Set the APN
+  request.set_apn(apn);
+
+  // Set the ue_state
+  config_ue_session_state(ue_state, request);
+
+  auto local_response = new AsyncLocalResponse<UESessionContextResponse>(
+      std::move(callback), RESPONSE_TIMEOUT);
+
+  auto response_reader = client.stub_->AsyncUpdateUEState(
+      local_response->get_context(), request, &client.queue_);
+  local_response->set_response_reader(std::move(response_reader));
+  return RETURNok;
+}
+
+//                    ADD : v4 with flow_dl
+//--------------------------------------------------
+int PipelinedServiceClient::UpdateUEIPv4SessionSetWithFlowdl(
```

Comment:
Same comment as above about moving the conversion logic into the other file.

Reply:
Comments addressed.